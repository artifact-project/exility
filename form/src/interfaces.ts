import Block from '@exility/block';

export interface ValueBox {
	name?: string;
	value?: string;
	checked?: boolean;
	selectedIndex?: number;
}

export interface Validity {
	id: string;
	detail: object & {value?: string, checked?: boolean, length?: number};
	nested?: Validity;
}

export type ValidateRule = (vbox: ValueBox) => Validity;

export interface UIErrorAttrs {
	for: string;
}

export interface UIFormAttrs {
	name?: string;
	context?: IFormContext;
}

export interface UIFormContext {
	$form: IFormContext;
}

export interface IUIElement {
	attrs: UIElementAttrs;
}

export interface IFormElement {
	name: string;
	value: string;
	checked: boolean;
	errors: {
		[name: string]: Validity;
	};
	invalid: boolean;
}

export interface IFormContext {
	id: string;
	handleEvent(block: IUIElement, event: Event);
	getBlockByDOMEvent(originalEvent: Event): typeof Block;

	connectForm(ui: Block<any, any>): void;
	disconnectForm(ui: Block<any, any>): void;

	connectElement(block: IUIElement): IFormElement;
	disconnectElement(block: IUIElement): IFormElement;

	connectError(ui: Block<any, any>): void;
	disconnectError(ui: Block<any, any>): void;
}

export type FormContextConfig = {
	validation?: {
		[name: string]: ValidateRule;
	};
}

export interface UIElementAttrs {
	name: string;
	type?: string;
	value?: string;
	checked?: boolean;
	required?: boolean;
	readOnly?: boolean;
	disabled?: boolean
	autoFocus?: boolean
	minLength?: number;
	maxLength?: number;
	tabIndex?: number;
	options?: ElementOptionAttrs[];
}

export interface ElementOptionAttrs {
	text: string;
	value?: string;
}

export interface UIElementContext {
	$form: IFormContext;
}
