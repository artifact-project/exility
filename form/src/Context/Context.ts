import {SimpleDict, T} from '@artifact-project/i18n';
import Block from '@exility/block';


import {FormContextConfig, IFormContext, IFormElement, ValidateRule, Validity, ValueBox} from '../interfaces';
import {minLength, createSomeOf, required} from '../rules/rules';
import UIElement from '../ui/Element/Element';
import {debounce, F_IMPORTANT, F_NO_ARGS} from '@perf-tools/balancer';
import UIForm from '../ui/Form/Form';
import UIError from '../ui/Error/Error';

const RADIO_TYPE = 'radio';
const CHECKBOX_TYPE = 'checkbox';
const SELECT_TYPE = 'select';

const FOCUS_EVENT_NAME = 'focus';
const BLUR_EVENT_NAME = 'blur';
const INPUT_EVENT_NAME = 'input';
const CHANGE_EVENT_NAME = 'change';

const EMPTY_ERRORS = {};

export class Element implements IFormElement {
	initialValue: string = '';
	initialChecked: boolean = false;
	initialSelectedIndex: number = 0;

	value: string = '';
	checked: boolean = false;
	selectedIndex: number = 0;

	get type(): string {
		return this.block.attrs.type;
	}

	get name(): string {
		return this.block.attrs.name;
	}

	get required(): boolean {
		return this.block.attrs.required;
	}

	get readOnly(): boolean {
		return this.block.attrs.readOnly || this.form.locked;
	}

	get disabled(): boolean {
		return this.block.attrs.disabled;
	}

	get minLength(): number {
		return +this.block.attrs.minLength || -1;
	}

	get maxLength(): number {
		return +this.block.attrs.maxLength || -1;
	}

	errors: {[name: string]: Validity} = {};
	active: boolean = false;
	changed: boolean = false;
	invalid: boolean = false;
	touched: boolean = false;
	validity: Validity = null;

	private validateRule: ValidateRule;

	constructor(public form: FormContext, public block: UIElement) {
		const rules: ValidateRule[] = form.getValidateRule(this.name);

		if (this.required) {
			rules.push(required());
		}

		if (this.minLength >= 0) {
			rules.push(minLength(this.minLength));
		}

		this.validateRule = createSomeOf(rules);
	}

	validate(vbox: ValueBox): Validity {
		return this.validateRule(vbox);
	}

	forceUpdate() {
		this.block.forceUpdate();
	}
}

export class ElementsGroup {
	private list: Element[] = [];

	get validaty() {
		return this.length ? this.list[0].validity : null;
	}

	get active() {
		return this.hasFlag('active');
	}

	get invalid() {
		return this.hasFlag('invalid');
	}

	get touched() {
		return this.hasFlag('touched');
	}

	get changed() {
		return this.hasFlag('changed');
	}

	get length() {
		return this.list.length;
	}

	constructor() {
	}

	eq(index: number) {
		return this.list[index];
	}

	hasFlag(name: string): boolean {
		let idx = this.length;

		if (idx) {
			if (idx === 1) {
				return this.list[0][name];
			}

			while (idx--) {
				if (this.list[idx][name]) {
					return true;
				}
			}
		}

		return false;
	}

	forEach(iterator: (el: Element, idx: number) => void, thisArg?: {}) {
		this.list.forEach(iterator, thisArg);
	}

	add(element: Element) {
		this.list.push(element);
	}

	remove(element: Element): boolean {
		const idx = this.list.indexOf(element);

		if (idx > -1) {
			this.list.splice(idx, 1);
			return true;
		} else {
			return false
		}
	}

	getErrorText(i18nDict?: SimpleDict) {
		if (this.length) {
			const validity = this.list[0].validity;
			return T(validity, i18nDict);
		}
	}
}

export class FormContext {
	id: string;
	values: {} = {};
	changed: boolean = false;
	invalid: boolean = false;
	locked: boolean = false;

	submitting: boolean = false;
	submitFailed: boolean = false;
	submitSucceeded: boolean = false;

	private forms: UIForm[] = [];
	private errors: UIError[] = [];

	private elementsGroups: {[name:string]: ElementsGroup} = {};
	private elementsIndex: {[cid:string]: Element} = {};

	private validateLock: boolean = false;
	private validateRules: any = {};

	constructor(private initialValues: object = {}, private config: FormContextConfig = {}) {
		this.validate = <() => void>debounce(
			this.validate,
			this,
			[],
			{flags: F_NO_ARGS & F_IMPORTANT},
		);
	}

	lock() {
		this.locked = true;
		this.forceUpdateAll();
	}

	unlock() {
		this.locked = false;
		this.forceUpdateAll();
	}

	getValidateRule(name: string) {
		const {validation = {}} = this.config;
		return validation.hasOwnProperty(name) ? [validation[name]] : [];
	}

	getElementsGroup(name) {
		let group = this.elementsGroups[name];

		if (group === void 0) {
			this.elementsGroups[name] = group = new ElementsGroup();
		}

		return group;
	}

	connectForm(form: UIForm) {
		this.forms.push(form);
	}

	disconnectForm(form: UIForm) {
		this.forms.splice(this.forms.indexOf(form), 1);
	}

	connectError(error: UIError) {
		// this.getElementsGroup(error.attrs.for).connectError(error);
		this.errors.push(error);
	}

	disconnectError(error: UIError) {
		// this.getElementsGroup(error.attrs.for).disconnectError(error);
	}

	connectElement(block: UIElement): Element {
		const {type, name} = block.attrs;
		const element = new Element(this, block);
		const elementsGroup = this.getElementsGroup(name);
		let {value} = block.attrs;
		let values = this.initialValues[name];

		if (!Array.isArray(values)) {
			values = values == null ? [] : [values];
		}

		if (type === CHECKBOX_TYPE || type === RADIO_TYPE) {
			if (value == null) {
				value = 'on';
			}

			element.checked = values.includes(value) || (block.attrs.checked !== null && block.attrs.checked);
		} else {
			const initialValue = values[elementsGroup.length];

			if (initialValue != null) {
				value = initialValue;
			}

			if (type === SELECT_TYPE) {
				element.selectedIndex = block.attrs.options.findIndex(opt => {
					return (opt.value == null ? opt.text : opt.value) == value;
				});
			}
		}

		element.value = value = (value == null ? '' : value + '');
		element.initialValue = element.value;
		element.initialChecked = element.checked;
		element.initialSelectedIndex = element.selectedIndex;

		elementsGroup.add(element);

		if (type === CHECKBOX_TYPE) {
			!this.values.hasOwnProperty(name) && (this.values[name] = []);
			element.checked && this.values[name].push(value);
		} else if (type === RADIO_TYPE) {
			!this.values.hasOwnProperty(name) && (this.values[name] = '');
			element.checked && (this.values[name] = value);
		} else {
			this.values[name] = value;
		}

		this.elementsIndex[block.cid] = element;
		this.validate();

		return element;
	}

	disconnectElement(block: UIElement) {
		const element = this.elementsIndex[block.cid];
		const elementsGroup = this.elementsGroups[block.attrs.name];

		if (elementsGroup && elementsGroup.remove(element)) {
			delete this.elementsIndex[block.cid];
		} else {
			throw new Error('[@exility/form] Unregister unknown block');
		}

		this.validate();
	}

	// todo: Этого метода не должно быть
	// возможно стоит сделать переменную `Block.activeBlock`, которую отслеживать по `focus`
	getBlockByDOMEvent(evt: Event) {
		const el = evt.target as HTMLInputElement;
		const elementsGroup = this.elementsGroups[el.name];
		const length = elementsGroup !== void 0 ? elementsGroup.length : 0;

		if (length === 0) {
			throw new Error('FAILED');
		} else if (length === 1) {
			return elementsGroup.eq(0).block;
		} else {
			return elementsGroup.eq(0).block;
		}
	}

	handleEvent(block: UIElement, evt: Event) {
		const type = evt.type;
		const target = evt.target as (HTMLInputElement & HTMLSelectElement);
		const element = this.elementsIndex[block.cid];
		const elementsGroup = this.elementsGroups[block.attrs.name];

		if (elementsGroup === void 0 || elementsGroup.length === 0) {
			console.warn('[@exility/form] Finding unregister element');
			return;
		}

		if (type === FOCUS_EVENT_NAME) {
			element.active = true;
		} else if (type === BLUR_EVENT_NAME) {
			element.active = false;
			element.touched = true;
		} else if (type === INPUT_EVENT_NAME || type === CHANGE_EVENT_NAME) {
			switch (element.type) {
				case SELECT_TYPE:
					element.value = target.value;
					element.selectedIndex = target.selectedIndex;
					break;

				case CHECKBOX_TYPE:
					element.checked = target.checked;
					break;

				case RADIO_TYPE:
					elementsGroup.forEach(el => {
						if (el !== element) {
							el.checked = false;
							el.changed = el.initialChecked !== el.changed;
							el.forceUpdate();
						}
					});
					element.checked = target.checked;
					break;

				default:
					element.value = target.value;
					break;
			}

			element.changed = (
				element.initialValue !== element.value ||
				element.initialChecked !== element.checked ||
				element.initialSelectedIndex !== element.selectedIndex
			);
		}

		element.forceUpdate();
		this.validate();
	}

	private validate() {
		const names = []; // названия элементов
		const targets: Element[] = []; // цели валидации

		this.changed = false;

		// Значения вормы
		const values = Object.keys(this.elementsIndex).reduce((values, key) => {
			const element = this.elementsIndex[key];
			const {type, name, value, checked} = element;

			targets.push(element);

			if (element.changed) {
				this.changed = true;
			}

			if (values.hasOwnProperty(name)) {
				if (type === CHECKBOX_TYPE) {
					checked && values[name].push(value);
				} else if (type === RADIO_TYPE) {
					checked && (values[name] = value);
				} else {
					values[name] = [].concat(values[name], value);
				}
			} else {
				names.push(name);

				if (type === CHECKBOX_TYPE) {
					values[name] = checked ? [value] : [];
				} else {
					values[name] = (type !== RADIO_TYPE || checked) ? value : '';
				}
			}

			return values;
		}, {});

		this.invalid = false;

		// Валидируем все цели
		targets.forEach((element) => {
			const curInvalidState = element.invalid;

			element.errors = EMPTY_ERRORS;
			element.invalid = false;

			const vbox = {
				name: element.name,
				value: element.value,
				checked: element.checked,
				values,
			};
			const validity = element.validate(vbox);

			if (isThenable(validity)) {
				// todo: короче, это надо потом делать, пока лень
				// todo: Незабыть про таймауты при изменении invalid и forceUpdate
				validity.then((validity) => {
					setValidation(this, element, validity, curInvalidState);
				});
			} else {
				setValidation(this, element, validity, curInvalidState);
			}
		});

		this.values = values;
		this.validateLock = false;

		forceUpdateAll(this.forms);
		forceUpdateAll(this.errors);
	}

	private forceUpdateAll() {
		forceUpdateAll(Object.values(this.elementsIndex));
	}
}

function forceUpdateAll(blocks: {forceUpdate(): void}[]) {
	switch (blocks.length) {
		case 1: forceUpdate(blocks[0]); break;
		default: blocks.forEach(forceUpdate);
	}
}

function forceUpdate(block: {forceUpdate(): void}) {
	block.forceUpdate();
}

function setValidation(form: FormContext, element: Element, validity: Validity, curInvalidState :boolean) {
	if (validity != null) {
		form.invalid = true;
		element.errors = {[validity.id]: validity};
		element.invalid = true;
		element.validity = validity;
	}

	(curInvalidState !== element.invalid) && element.forceUpdate();
}


function isThenable(validity: {then?: Function}): validity is Promise<Validity> {
	return !!(validity && validity.then);
}
