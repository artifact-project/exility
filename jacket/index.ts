import mount, {DOMWrapper} from './src/mount/mount';

export {
	mount,
	DOMWrapper,
};
