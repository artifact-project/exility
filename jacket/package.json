{
  "name": "@exility/jacket",
  "version": "0.1.0",
  "description": "Exility Jacket is a testing utility for your blocks.",
  "main": "index.js",
  "scripts": {
    "test": "jest --coverage",
    "ts-build": "tsc",
    "git-clean": "git clean -f",
    "prepublish": "npm run test && npm run ts-build",
    "postpublish": "npm run git-clean"
  },
  "files": [
    "*.js",
    "*.d.ts",
    "src/*/*.js",
    "src/*/*.d.ts"
  ],
  "author": "RubaXa <ibnRubaXa@gmail.com>",
  "license": "MIT",
  "dependencies": {
    "tslib": ">=1.8.1"
  },
  "devDependencies": {
    "@types/jest": "^19.2.3",
    "jest": "^20.0.3",
    "ts-jest": "^20.0.4",
    "typescript": "^2.6.1",
    "@exility/ts-transformer": "*",
    "@exility/block": "*",
    "@exility/dom": "*"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "<rootDir>/__jest__/preprocessor.js"
    },
    "testRegex": "\\.tests\\.ts$",
    "setupTestFrameworkScriptFile": "<rootDir>/__jest__/extensions.ts",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "json"
    ]
  }
}
