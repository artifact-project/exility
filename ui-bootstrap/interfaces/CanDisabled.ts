import regression from './_regression';

interface CanDisabled {
	disabled?: boolean;
}

export default CanDisabled;

regression<CanDisabled>('CanDisabled', {
	'disabled: undefined -> true': {
		initialData: {},
		initialTest(wrapper) {
			expect(wrapper.attr('disabled')).toBe(false);
			expect(wrapper.attr('aria-disabled')).toBe(null);
		},

		data: {disabled: true},
		test(wrapper) {
			expect(wrapper.attr('disabled')).toBe(true);
			expect(wrapper.attr('aria-disabled')).toBe('true');
		},
	},

	'disabled: true -> false': {
		initialData: {disabled: true},
		initialTest(wrapper) {
			expect(wrapper.attr('disabled')).toBe(true);
			expect(wrapper.attr('aria-disabled')).toBe('true');
		},

		data: {disabled: false},
		test(wrapper) {
			expect(wrapper.attr('disabled')).toBe(false);
			expect(wrapper.attr('aria-disabled')).toBe(null);
		},
	},
});
