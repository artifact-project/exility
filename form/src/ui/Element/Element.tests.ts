import {create} from '@exility/jacket';
import UIElement from './Element';
import {FormContext} from '../../Context/Context';
import {requestFrame} from '@perf-tools/balancer';
import {UIElementAttrs, UIElementContext} from '../../interfaces';

describe('ui / Element', () => {
	let context;

	async function frame() {
		return new Promise(resolve => {
			requestFrame(resolve);
		});
	}

	async function simulateEvent(wrapper, event = 'input') {
		context.$form.handleEvent(wrapper.target, {
			type: event,
			target: wrapper.target.getRootNode(),
		});
		await frame();
	}

	beforeEach(() => {
		context = {
			$form: new FormContext({
				inline: 'Wow!',
				array: ['foo'],
				switcher: 'bar',
				color: 'blue',
			}, {
				rules: {
					// Only English alphabet
					username: ({value}) => /^[a-z]*$/i.test(value) ? null : {
						id: 'onlyEng',
						detail: null,
					},
				},
				submit: () => Promise.resolve(true),
			}),
		};
	});

	describe('text', () => {
		it('empty', async () => {
			const text = create<UIElementAttrs, UIElementContext>(
				UIElement,
				{name: 'empty'},
				context,
			);

			expect(text).toMatchSnapshot();
			expect((text.target as any).input instanceof HTMLInputElement).toBe(true);
			expect(context.$form.values).toEqual({empty: ''});

			text.attr('value', 'ok');
			await simulateEvent(text);

			expect(context.$form.values).toEqual({empty: 'ok'});
		});

		it('with initial value', async () => {
			const text = create(
				UIElement,
				{name: 'empty', value: 'Foo'},
				context,
			);
			expect(text).toMatchSnapshot();
			expect(context.$form.values).toEqual({empty: 'Foo'});
		});

		it('with value', async () => {
			expect(create<UIElementAttrs, UIElementContext>(
				UIElement,
				{name: 'inline'},
				context,
			)).toMatchSnapshot();
		});
	});

	describe('checkbox', () => {
		it('empty', async () => {
			const cbx = create(
				UIElement,
				{name: 'empty', type: 'checkbox'},
				context,
			);

			expect(cbx).toMatchSnapshot();
			expect(cbx.attr('checked')).toBe(false);
		});

		it('with inline checked', async () => {
			const cbx = create(
				UIElement,
				{name: 'empty', type: 'checkbox', checked: true},
				context,
			);

			expect(cbx.attr('checked')).toBe(true);
		});

		it('with checked: foo', async () => {
			const cbx = create(
				UIElement,
				{name: 'array', type: 'checkbox', value: 'foo'},
				context,
			);

			expect(cbx).toMatchSnapshot();
			expect(cbx.attr('checked')).toBe(true);
		});

		it('with checked: bar', async () => {
			const foo = create(
				UIElement,
				{name: 'array', type: 'checkbox', value: 'foo'},
				context,
			);

			const bar = create(
				UIElement,
				{name: 'array', type: 'checkbox', value: 'bar'},
				context,
			);

			expect(bar).toMatchSnapshot();
			expect(context.$form.values).toEqual({array: ['foo']});
			expect(foo.attr('checked')).toBe(true);
			expect(bar.attr('checked')).toBe(false);

			bar.attr('checked', true);
			await simulateEvent(bar);
			expect(context.$form.values).toEqual({array: ['foo', 'bar']});

			foo.attr('checked', false);
			await simulateEvent(foo);
			expect(context.$form.values).toEqual({array: ['bar']});

			bar.attr('checked', false);
			await simulateEvent(bar);
			expect(context.$form.values).toEqual({array: []});
		});
	});

	describe('radio', () => {
		it('empty', async () => {
			const radio = create(
				UIElement,
				{name: 'empty', type: 'radio'},
				context,
			);

			expect(radio).toMatchSnapshot();
			expect(radio.attr('checked')).toBe(false);
		});

		it('with checked', async () => {
			const foo = create(
				UIElement,
				{name: 'switcher', type: 'radio', value: 'foo'},
				context,
			);

			const bar = create(
				UIElement,
				{name: 'switcher', type: 'radio', value: 'bar'},
				context,
			);

			const baz = create(
				UIElement,
				{name: 'switcher', type: 'radio', value: 'baz'},
				context,
			);

			expect(foo.attr('checked')).toBe(false);
			expect(bar.attr('checked')).toBe(true);
			expect(baz.attr('checked')).toBe(false);
			expect(context.$form.values).toEqual({switcher: 'bar'});

			baz.attr('checked', true);
			await simulateEvent(baz);

			expect(foo.attr('checked')).toBe(false);
			expect(bar.attr('checked')).toBe(false);
			expect(baz.attr('checked')).toBe(true);
			expect(context.$form.values).toEqual({switcher: 'baz'});
		});
	});

	describe('textarea', () => {
		it('with initial value', async () => {
			expect(create(
				UIElement,
				{name: 'empty', value: 'Foo', type: 'textarea'},
				context,
			)).toMatchSnapshot();
		});
	});

	describe('select', () => {
		it('with value', async () => {
			const select = create(
				UIElement,
				{
					name: 'color',
					type: 'select',
					options: [
						{value: 'red', text: 'Red'},
						{value: 'green', text: 'Green'},
						{value: 'blue', text: 'Blue'},
					],
				},
				context,
			);

			expect(select).toMatchSnapshot();
			expect(select.attr('selectedIndex')).toBe(2);
		});
	});

	it('focus/blur', async () => {
		const text = create<UIElementAttrs, UIElementContext>(
			UIElement,
			{name: 'empty'},
			context,
		);

		await simulateEvent(text, 'focus');
		expect(text.classList).toEqual(['active', 'is-text']);

		await simulateEvent(text, 'blur');
		expect(text.classList).toEqual(['is-text', 'touched']);
	});

	it('changed', async () => {
		const text = create<UIElementAttrs, UIElementContext>(
			UIElement,
			{name: 'empty'},
			context,
		);

		await simulateEvent(text, 'focus');
		text.attr('value', 'wow');
		await simulateEvent(text);

		expect(text.classList).toEqual(['active', 'changed', 'is-text']);

		await simulateEvent(text, 'blur');
		expect(text.classList).toEqual(['changed', 'is-text', 'touched']);

		text.attr('value', '');
		await simulateEvent(text);
		expect(text.classList).toEqual(['is-text', 'touched']);
	});

	it('invalid / required', async () => {
		const text = create(
			UIElement,
			{name: 'username', required: true},
			context,
		);
		const textFormElem = (text.target as UIElement).getFormElement();

		await frame();
		expect(!!textFormElem.errors).toBe(true);
		expect(text.classList).toEqual(['invalid', 'is-text']);

		text.attr('value', 'X');
		await simulateEvent(text);
		expect(text.classList).toEqual(['changed', 'is-text']);
		expect(textFormElem.errors).toEqual({});
	});

	it('invalid / minLength', async () => {
		const text = create(
			UIElement,
			{name: 'username', minLength: 3},
			context,
		);
		const textFormElem = (text.target as UIElement).getFormElement();

		text.attr('value', 'X');
		await simulateEvent(text);

		expect(!!textFormElem.errors.minLength).toBe(true);
		expect(text.classList).toEqual(['changed', 'invalid', 'is-text']);

		text.attr('value', 'Xy!');
		await simulateEvent(text);
		expect(!!textFormElem.errors.minLength).toBe(false);
		expect(!!textFormElem.errors.onlyEng).toBe(true);
		expect(text.classList).toEqual(['changed', 'invalid', 'is-text']);

		text.attr('value', 'Xyz');
		await simulateEvent(text);
		expect(text.classList).toEqual(['changed', 'is-text']);
		expect(textFormElem.errors).toEqual({});
	});

	it('readOnly/LockedForm', async () => {
		const text = create<UIElementAttrs, UIElementContext>(
			UIElement,
			{name: 'empty'},
			context,
		);

		context.$form.lock();
		expect(text.attr('readOnly')).toBe(true);
		expect(text.classList).toEqual(['is-text', 'readOnly']);

		context.$form.unlock();
		expect(text.attr('readOnly')).toBe(false);
		expect(text.classList).toEqual(['is-text']);
	});
});
