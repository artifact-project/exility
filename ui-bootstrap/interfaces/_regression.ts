import {registerRegression, RegressionEnv, Regression, RegressionCases} from 'ibi';
import {mount, DOMWrapper} from '@exility/jacket';

export default function regression<T>(name: string, cases: RegressionCases<DOMWrapper, T, RegressionEnv>) {
	class UIRegression extends Regression<DOMWrapper, T, RegressionEnv> {
		factory(Class, data) {
			return mount(new Class(data));
		}

		update(wrapper, data) {
			wrapper.update(data);
		}

		getAll() {
			return cases;
		}
	}

	registerRegression(name, UIRegression);
}
