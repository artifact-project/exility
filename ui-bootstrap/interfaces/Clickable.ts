import regression from './_regression';
import CanDisabled from './CanDisabled';

interface Clickable extends CanDisabled {
	'@click'?(evt: Event): void;
}

export default Clickable;

regression<Clickable>('Clickable', {
	'onClick': {
		initialData: {},
		initialTest(wrapper) {
			const log: MouseEvent[] = [];

			wrapper.on<MouseEvent>('click', (evt) => log.push(evt));
			wrapper.simulate('click');

			expect(log.length).toBe(1);
			expect(log[0].type).toBe('click');
		},

		data: {disabled: true},
		test(wrapper) {
			const log: MouseEvent[] = [];

			wrapper.on<MouseEvent>('click', (evt) => log.push(evt));
			wrapper.simulate('click');

			expect(log.length).toBe(0);
		},
	},
});
