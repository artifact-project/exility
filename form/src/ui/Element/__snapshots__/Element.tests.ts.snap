// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui / Element checkbox empty 1`] = `
<input class="is-checkbox"
       name="empty"
       type="checkbox"
       value="on"
>
`;

exports[`ui / Element checkbox with checked: bar 1`] = `
<input class="is-checkbox"
       name="array"
       type="checkbox"
       value="bar"
>
`;

exports[`ui / Element checkbox with checked: foo 1`] = `
<input class="is-checkbox"
       name="array"
       type="checkbox"
       value="foo"
>
`;

exports[`ui / Element radio empty 1`] = `
<input class="is-radio"
       name="empty"
       type="radio"
       value="on"
>
`;

exports[`ui / Element select with value 1`] = `
<select class="is-select"
        name="color"
        type="select"
        value="blue"
>
  <option value="red">
    Red
  </option>
  <option value="green">
    Green
  </option>
  <option value="blue">
    Blue
  </option>
</select>
`;

exports[`ui / Element text empty 1`] = `
<input class="is-text"
       name="empty"
       type="text"
       value
>
`;

exports[`ui / Element text with initial value 1`] = `
<input class="is-text"
       name="empty"
       type="text"
       value="Foo"
>
`;

exports[`ui / Element text with value 1`] = `
<input class="is-text"
       name="inline"
       type="text"
       value="Wow!"
>
`;

exports[`ui / Element textarea with initial value 1`] = `
<textarea class="is-textarea"
          name="empty"
          type="textarea"
          value="Foo"
>
</textarea>
`;
